src/jdk.jlink/share/classes/jdk/tools/jlink/internal/Snippets.java

```diff
@@ -0,0 +1,461 @@
+     * overloading the 64k code limit.
+     */
+    public static abstract class CollectionSnippetBuilder {
+        /**
+         * Tested page size of string array
+         */
+        public static final int STRING_PAGE_SIZE = 8000;
+
+        /**
+         * Tested page size of enum array
+         */
+        public static final int ENUM_PAGE_SIZE = 5000;
+
+        /**
+         * Good enough for average ~30 bytes per element
+         */
+        public static final int DEFAULT_PAGE_SIZE = 2000;
+
+        /**
+         * Arbitary default values based on 15K code size on ~30 bytes per element
+         */
+        protected static final int DEFAULT_THRESHOLD = 512;
+
+        protected ClassDesc elementType;
+        protected ClassDesc ownerClassDesc;
+        protected ClassBuilder clb;
+
+        // Default values enable pagination by default
+        protected String methodNamePrefix = "csb" + Integer.toHexString(hashCode()) + "Page";
+        protected int activatePagingThreshold = DEFAULT_THRESHOLD;
+        protected int pageSize = DEFAULT_PAGE_SIZE;
+
+        /**
+         * @param elementType The element type
+         */
+        protected CollectionSnippetBuilder(ClassDesc elementType) {
+            this.elementType = Objects.requireNonNull(elementType);
+        }
+
+        /**
+         * Enable pagination if the count of elements is larger than the given threshold.
+         *
+         * @param methodNamePrefix The method name prefix for generated paging helper methods
+         * @param pageSize         The page size
+         * @param threshold        The element count to actiave the pagination
+         */
+        public CollectionSnippetBuilder enablePagination(String methodNamePrefix, int pageSize, int threshold) {
+            return this.pageSize(pageSize)
+                       .activatePagingThreshold(threshold)
+                       .methodNamePrefix(methodNamePrefix);
+        }
+
+        /**
+         * Enable pagination if the count of elements is larger than pageSize or DEFAULT_THRESHOLD
+         */
+        public CollectionSnippetBuilder enablePagination(String methodNamePrefix, int pageSize) {
+            return enablePagination(methodNamePrefix, pageSize, Math.min(pageSize, DEFAULT_THRESHOLD));
+        }
+
+        /**
+         * Enable pagination if the count of elements is larger than pageSize or DEFAULT_THRESHOLD
+         * with page size DEFAULT_PAGE_SIZE.
+         */
+        public CollectionSnippetBuilder enablePagination(String methodNamePrefix) {
+            return enablePagination(methodNamePrefix, DEFAULT_PAGE_SIZE, DEFAULT_THRESHOLD);
+        }
+
+        /**
+         * Disable pagination. Generated bytecode will always try to construct the collection inline.
+         */
+        public CollectionSnippetBuilder disablePagination() {
+            this.activatePagingThreshold = -1;
+            this.methodNamePrefix = null;
```

Comment:
suggest to reset `activatePagingThreshold` and `pageSize` to default value.   Set `methodNamePrefix` to null which indicates pagination is disabled.   So the constructor should initialize `methodNamePrefix` to null by default.